store/sqlstore/channel_store.go

```diff
@@ -518,27 +518,32 @@ func (s SqlChannelStore) Save(channel *model.Channel, maxChannelsPerTeam int64)
 		return nil, store.NewErrInvalidInput("Channel", "Type", channel.Type)
 	}
 
-	transaction, err := s.GetMaster().Begin()
-	if err != nil {
-		return nil, errors.Wrapf(err, "begin_transaction: ")
-	}
-	defer finalizeTransaction(transaction)
-
-	newChannel, err := s.saveChannelT(transaction, channel, maxChannelsPerTeam)
-	if err != nil {
-		return newChannel, err
-	}
+	var newChannel *model.Channel
+	err := store.WithDeadlockRetry(func() error {
+		transaction, err := s.GetMaster().Begin()
+		if err != nil {
+			return errors.Wrapf(err, "begin_transaction: ")
+		}
+		defer finalizeTransaction(transaction)
 
-	// Additionally propagate the write to the PublicChannels table.
-	if err := s.upsertPublicChannelT(transaction, newChannel); err != nil {
-		return nil, errors.Wrapf(err, "upsert_public_channel: ")
-	}
+		newChannel, err = s.saveChannelT(transaction, channel, maxChannelsPerTeam)
+		if err != nil {
+			return err
+		}
 
-	if err := transaction.Commit(); err != nil {
-		return nil, errors.Wrapf(err, "commit_transaction: ")
-	}
+		// Additionally propagate the write to the PublicChannels table.
+		if err := s.upsertPublicChannelT(transaction, newChannel); err != nil {
+			return errors.Wrapf(err, "upsert_public_channel: ")
+		}
 
-	return newChannel, nil
+		if err := transaction.Commit(); err != nil {
+			return errors.Wrapf(err, "commit_transaction: ")
+		}
+		return nil
+	})
+	// There are cases when in case of conflict, the original channel value is returned.
```

Comment:
I note that while we're now consistently returning `newChannel`, where there were cases before when we /didn't/. Is it worth fixing the caller in these cases to not rely on this behaviour?

Reply:
I'll do a sweep and check for any such cases.